components/Common/SearchBar/index.tsx

```diff
@@ -0,0 +1,337 @@
+import { MdTravelExplore, MdClose } from 'react-icons/md';
+import classNames from 'classnames';
+import { AnimatePresence, motion } from 'framer-motion';
+import Link from 'next/link';
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { FormattedMessage } from 'react-intl';
+import styles from './index.module.scss';
+import SectionTitle from '../SectionTitle';
+import { useClickOutside } from '../../../hooks/useClickOutside';
+import useKeyPress from '../../../hooks/useKeypress';
+import type { FC } from 'react';
+import type { SearchFunction, SearchResult } from '../../../types';
+
+type SearchBarProps = {
+  search?: SearchFunction;
+  setup?: () => Promise<void>;
+  footer?: FC | (() => JSX.Element);
+};
+
+const containerTransition = { type: 'spring', damping: 22, stiffness: 150 };
+
+const containerVariants = {
+  expanded: {
+    width: 'auto',
+    height: 'auto',
+  },
+  collapsed: {
+    width: 0,
+    height: '3em',
+  },
+};
+
+const SearchBar: FC<SearchBarProps> = ({
+  setup,
+  search,
+  footer: Footer,
+}: SearchBarProps) => {
+  const [isLoading, setIsLoading] = useState(true);
+  const searchInputRef = useRef<HTMLInputElement | null>(null);
+  const listRef = useRef<HTMLUListElement | null>(null);
+  const activeResultRef = useRef<number>(-1);
+
+  const [isExpanded, setExpanded] = useState(false);
+  const [query, setQuery] = useState('');
+  const [results, setResults] = useState<SearchResult[]>([]);
+  const isEmpty = !results || results.length === 0;
+
+  const containerClassNames = classNames(styles.searchBarContainer, {
+    [styles.expanded]: isExpanded,
+  });
+
+  const resultsClassNames = classNames(styles.searchResults, {
+    [styles.noResults]: isEmpty,
+  });
+
+  const expandContainer = useCallback(() => {
+    setExpanded(true);
+  }, [setExpanded]);
+
+  const collapseContainer = useCallback(() => {
+    setQuery('');
+    setExpanded(false);
+  }, [setQuery, setExpanded]);
+
+  const selectItem = useCallback(
+    (index: number) => {
+      if (!listRef.current) {
+        return;
+      }
+
+      const el = listRef.current.children[index]!;
+      el.querySelector('a')?.focus();
+      el.scrollIntoView({ behavior: 'smooth', block: 'nearest' });
+    },
+    [listRef]
+  );
+
+  const toggleContainer = useCallback(() => {
+    isExpanded ? collapseContainer() : expandContainer();
+  }, [isExpanded, collapseContainer, expandContainer]);
+
+  const onKeyPressHandler = useCallback(() => {
+    if (!isExpanded) {
+      expandContainer();
+    }
+  }, [isExpanded, expandContainer]);
+
+  const onQueryChangeHandler = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      e.preventDefault();
+
+      // Clear results when the query is emptied
+      if (!e.target.value) {
+        setResults([]);
+      }
+
+      setQuery(e.target.value);
+    },
+    [setQuery, setResults]
```

Comment:
```suggestion
    []
```
We don't need to add these as they will never change.

Reply:
Here and below: while I agree on that, I tend to pass all values (including immutable one) to useCallback/useMemo in order to have a quick view of what is used in the function. I don't think it affect performance or are there any cons in doing this I'm not aware of?